### ISSUE TITLE
Optax `opt_state` changes shape on first training iteration, leading to recompilation

### ISSUE THREAD

**Issue by i-m-mll** [mlprt/feedbax#4]
In `feedbax.trainer.TaskTrainer`, before starting the training loop, the training and validation methods are executed a single time in order to JIT compile them. 

https://github.com/mlprt/feedbax/blob/ed0d0e0ebc4cb25b51cc85742a11000430b4ffcf/feedbax/train.py#L259-L284

Note that wrapping this in a ` tqdm` loop was forquick way to time the compilation, though I ought to get rid of the trivial progress bar and just log the compilation time. 

Having compiled these functions, I'd expected the training loop to warm up very quickly. However, the first step of the training loop is also slow. I suspect this is because `train_loop` is being recompiled, due to `optimizer.update` changing the shape of `opt_state` -- but only on the first iteration.

This could be solved by keeping the result of the JIT compilation/keeping the first two training iterations outside of the main progress bar. However, I'd still like an answer about why `optimizer.update` changes the shape of `opt_state`. It kind of makes sense that it would do so if it initializes its state lazily, and changes the state structure when (for example) it first encounters gradients.
